File: elixir/flatten_array.ex
defmodule FlattenArray do
  @doc """
    Accept a list and return the list flattened without nil values.

    ## Examples

      iex> FlattenArray.flatten([1, [2], 3, nil])
      [1,2,3]

      iex> FlattenArray.flatten([nil, nil])
      []

  """
  @spec flatten(list) :: list
  def flatten([]), do: []

  def flatten([x | xs]) when is_list(x) do
    flatten(x) ++ flatten(xs)
  end

  def flatten([nil | xs]), do: flatten(xs)
  def flatten([x | xs]), do: [x | flatten(xs)]
end

File: prolog/flatten_array.pl
flatten_list([], []).
flatten_list([H|T], Flattened) :-
    H = nil,
    flatten_list(T, Flattened).
flatten_list([H|T], Flattened) :-
    is_list(H),
    flatten_list(H, Hs),
    flatten_list(T, Ts),
    append(Hs, Ts, Flattened).
flatten_list([H|T], Flattened) :-
    flatten_list(T, Ts),
    Flattened = [H|Ts].
